File: netCDF/hands_on/Solutions/F90/hands_on_compression.f90
program hands_on_compression 
  use netcdf
  implicit none
  integer :: ierr
  
  character (len = *), parameter :: FILE_NAME = "var2d_compression.nc"
  integer,parameter :: NDIMS=2
  integer,parameter :: NI_GLO=1000
  integer,parameter :: NJ_GLO=1000

  integer :: field_glo(NI_GLO,NJ_GLO)
  integer :: t
  integer :: ncid,lat_dimid,lon_dimid,varid
  integer :: shuffle, deflate, deflate_level
  integer :: dimids(NDIMS), starts(NDIMS), counts(NDIMS)

  ! Create the netcdf file.
  call check(nf90_create(FILE_NAME,NF90_NETCDF4,ncid))
  
  ! Define the dimensions. NetCDF will hand back an ID for each. 
  call check(nf90_def_dim(ncid, "lat", NI_GLO, lat_dimid))
  call check(nf90_def_dim(ncid, "lon", NJ_GLO, lon_dimid))

  ! Initialize global array.
  call init_global(field_glo)

  ! Define variable var and add an attribute.
  call check(nf90_def_var(ncid, "var_2d_temp", nf90_int, &
                      (/ lat_dimid, lon_dimid/), varid))
  call check(nf90_put_att(ncid, varid, "standart_name", "dummy_temporal_variable"))

  ! Define compression level
  shuffle = 1;
  deflate = 1;
  deflate_level = 0;
  call check(nf90_def_var_deflate(ncid, varid, shuffle, deflate, deflate_level))

  ! End define mode. This tells netCDF we are done defining metadata.
  call check(nf90_enddef(ncid)) 

  ! Write the variable.
  call check(nf90_put_var(ncid, varid, field_glo))

  ! Close file.
  call check(nf90_close(ncid))

  contains
  ! This subroutine comes with the netCDF package. 
  ! It handles errors by printing an error message  and exiting with a non-zero status.
  subroutine check(status)
    integer,intent(in) :: status
    if(status /= nf90_noerr) then 
      print *, trim(nf90_strerror(status))
      stop "Stopped"
    end if
  end subroutine check

  ! This subroutine fills in global arrays containing values of the variable, longitude, and latitude.
  subroutine init_global(field_glo)
  integer,intent(inout) :: field_glo(NI_GLO,NJ_GLO)        
  integer i,j
  
  do j=1,NJ_GLO
    do i=1,NI_GLO
      field_glo(i,j)=(i-1)+(j-1)*NI_GLO
    enddo
  enddo
  end subroutine init_global

end program hands_on_compression

File: netCDF/hands_on/Solutions/C/hands_on_compression.c
#include "mpi.h"
#include <netcdf.h>
#include <netcdf_par.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NDIMS     2
#define NI_GLO    1000 
#define NJ_GLO    1000

#define ERR(e) {printf("Error: %s\n", nc_strerror(e)); return 2;}

// Function fills in data
void fill_data(int* data, int ni, int nj)
{
  int i,j;
  for(i=0; i < ni; i++)
    for(j=0; j < nj; j++)
      data[j+i*nj] = j + i*nj;
}

int main (void)
{
  int mpirank, mpisize;
  /* Variables holding netCDF ids of the file and the variables stored in it */
  int ncid, varid;
  /* Variables holding dimensions: longitude, latitude  */
  int lon_dimid, lat_dimid, temp_dimid;
  /* Array holding dimensions of the 2D varialbe */
  int var_dimids[NDIMS];
  /* Compression variables */
  int shuffle, deflate, deflate_level;
  int err;
  int ni, nj, ibegin, jbegin;
  char filename[128] = "var2d_compression.nc";

  //--------------------------------------
  //  NetCDF metadata definitions
  //--------------------------------------

  /* Create a netcdf file */
  if (err=nc_create(filename, NC_NETCDF4, &ncid))
    ERR(err);

  /* Define two spatial dimensions */
  if (err=nc_def_dim(ncid, "lat", NI_GLO, &lat_dimid))
    ERR(err);
  if (err=nc_def_dim(ncid, "lon", NJ_GLO, &lon_dimid))
    ERR(err);

  /* Define a 2D temporal variable and add an attribute */
  var_dimids[0] = lat_dimid;
  var_dimids[1] = lon_dimid;
  if (err=nc_def_var(ncid, "var_2d", NC_INT, NDIMS, var_dimids, &varid))
    ERR(err);
  if (err=nc_put_att(ncid, varid, "standart_name", NC_CHAR, strlen("dummy_temporal_variable"), "dummy_temporal_variable"))
    ERR(err);

  /* Define compression level*/
  shuffle = 1;
  deflate = 1;
  deflate_level = 9;
  if (err=nc_def_var_deflate(ncid, varid, shuffle, deflate, deflate_level))
    ERR(err);

  /* Close metadata definition */
  if (err=nc_enddef(ncid))
    ERR(err);

  ni=NI_GLO; nj=NJ_GLO;
  /* Fill in data */
  int *data = (int *)malloc(sizeof(int)*ni*nj);
  fill_data(data, ni, nj);

  /* Write data */
  if (err=nc_put_var(ncid, varid, data)) ERR(err);

  /* Close the file */
  if (err=nc_close(ncid)) ERR(err);

  free(data);

  return 0;
}     

